Feature/add trailer video to movie and tvshow

Noor-Serry opened the pull request:
# Pull Request Template

## Description
add trailer video to movie and tvshow 
---

## Changes Made
List the changes introduced in this pull request:

- movie && tv show api service
- dtos
- mappers
-- add function in ui to open the video
---

## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.


https://github.com/user-attachments/assets/2a955646-a554-43c7-923a-a846c8970414


---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

Noor-Serry committed:
feature: implement dynamic launcher icons for light/dark mode

This commit introduces dynamic launcher icons that change based on the system's light or dark mode.

Key changes:
- Added `LightLauncher` and `DarkLauncher` activity aliases in `AndroidManifest.xml`.
- The `LightLauncher` is enabled by default.
- Implemented logic in `MainActivity` to switch between launcher icons in `onDestroy` based on the current UI mode (light/dark).
- The `switchLauncherIcon` function enables the appropriate alias and disables the other.
- Removed the `IconManager.kt` file as its functionality is now handled within `MainActivity`.
- Updated `MainActivity` in `AndroidManifest.xml` to `enabled="false"` as the aliases will be used for launching.
- Added `configChanges="uiMode"` to `MainActivity` to handle UI mode changes.
- Set `taskAffinity=""` and `launchMode="singleTop"` for `MainActivity` in `AndroidManifest.xml`.
- Specified the package name in `AndroidManifest.xml`.

Noor-Serry committed:
feature: implement play movie video

This commit introduces the ability to play movie videos.

- Added a `videoUrl` field to the `Movie` and `TvShow` entities.
- Updated the `MovieRemoteMapper` and `MovieDetailRemoteMapper` to populate the `videoUrl`.
- Modified the `PlayButton` to be clickable and trigger an `onClick` callback.
- In `MovieDetailsScreen`, the `PlayButton` now calls `interactionListener::onPlayVideoClicked` when clicked.
- Added `LaunchMovieVideoEffect` to `MovieDetailsEffect` to handle video launching.
- Implemented `onPlayVideoClicked` in `MovieDetailsViewModel` to send the `LaunchMovieVideoEffect`.
- Created `openYouTubeVideo` function in `ui/screens/VideoLauncher.kt` to open YouTube videos either in the app or in a web browser.
- Removed the `hasVideo` field from `LocalMovieDto` and related mappers as `videoUrl` provides this information.
- The `MovieApiService` now includes `include_video_language` parameter in the `getMovieDetailsById` call.
- Added new DTOs for video responses: `VideoResponse`, `VideoResultsWrapperDto`, and `VideoDto`.
- Added `VideoBaseUrl` object for YouTube base URL.

Noor-Serry committed:
feature: implement play movie video

This commit introduces the ability to play movie videos.

- Added a `videoUrl` field to the `Movie` and `TvShow` entities.
- Updated the `MovieRemoteMapper` and `MovieDetailRemoteMapper` to populate the `videoUrl`.
- Modified the `PlayButton` to be clickable and trigger an `onClick` callback.
- In `MovieDetailsScreen`, the `PlayButton` now calls `interactionListener::onPlayVideoClicked` when clicked.
- Added `LaunchMovieVideoEffect` to `MovieDetailsEffect` to handle video launching.
- Implemented `onPlayVideoClicked` in `MovieDetailsViewModel` to send the `LaunchMovieVideoEffect`.
- Created `openYouTubeVideo` function in `ui/screens/VideoLauncher.kt` to open YouTube videos either in the app or in a web browser.
- Removed the `hasVideo` field from `LocalMovieDto` and related mappers as `videoUrl` provides this information.
- The `MovieApiService` now includes `include_video_language` parameter in the `getMovieDetailsById` call.
- Added new DTOs for video responses: `VideoResponse`, `VideoResultsWrapperDto`, and `VideoDto`.
- Added `VideoBaseUrl` object for YouTube base URL.

Noor-Serry committed:
Merge remote-tracking branch 'origin/feature/add-trailer-video-to-movie-and-tvshow' into feature/add-trailer-video-to-movie-and-tvshow

monaayman123 committed:
feature: implement play episode video

monaayman123 committed:
feature: implement play tv show video

monaayman123 committed:
feature: implement play tv show episode video and

Noor-Serry committed:
refactor: Add fullVideoUrl property to VideoDto

This commit introduces a `fullVideoUrl` property to the `VideoDto` class.
This property constructs the complete YouTube video URL by appending the video `key` to the `YOUTUBE_BASE_URL`.

The `MovieDetailRemoteMapper` has been updated to utilize this new property, simplifying the video URL retrieval logic.

Noor-Serry committed:
Merge remote-tracking branch 'origin/feature/add-trailer-video-to-movie-and-tvshow' into feature/add-trailer-video-to-movie-and-tvshow

Noor-Serry committed:
refactor: remove video URLs from episode data

This commit removes the video URL field from the `Episode` entity, `EpisodeDto`, `EpisodeUiState`, and related mappers.

The following changes were made:
- Removed `videoUrl` from `Episode.kt`, `EpisodeDto.kt`, `EpisodeUiState.kt`.
- Updated `EpisodeRemoteMapper.kt` and `SeriesDetailsStateMapper.kt` to reflect the removal of `videoUrl`.
- Removed `onPlayEpisodeClicked` from `SeriesDetailsInteractionListener.kt` and its implementation in `SeriesDetailsViewModel.kt` and `SeriesDetailsScreen.kt`.
- Removed `findVideoUrlForEpisode` from `SeriesDetailsViewModel.kt`.
- Modified `TvShowsApiService.kt` to remove the `include_video_language` parameter from `getEpisodesBySeasonNumber` as video URLs are no longer being fetched per episode.
- Updated `VideoDto.kt` to remove unused import.
- Updated `TvShowDetailsRemoteMapper.kt` to use `fullVideoUrl` from `VideoDto`.
- Updated `AndroidManifest.xml` to simplify launcher activity declarations and remove the light/dark theme-specific aliases.
- Added logging in `VideoLauncher.kt` for debugging.

Noor-Serry committed:
refactor: remove video URLs from episode data

This commit removes the video URL field from the `Episode` entity, `EpisodeDto`, `EpisodeUiState`, and related mappers.

The following changes were made:
- Removed `videoUrl` from `Episode.kt`, `EpisodeDto.kt`, `EpisodeUiState.kt`.
- Updated `EpisodeRemoteMapper.kt` and `SeriesDetailsStateMapper.kt` to reflect the removal of `videoUrl`.
- Removed `onPlayEpisodeClicked` from `SeriesDetailsInteractionListener.kt` and its implementation in `SeriesDetailsViewModel.kt` and `SeriesDetailsScreen.kt`.
- Removed `findVideoUrlForEpisode` from `SeriesDetailsViewModel.kt`.
- Modified `TvShowsApiService.kt` to remove the `include_video_language` parameter from `getEpisodesBySeasonNumber` as video URLs are no longer being fetched per episode.
- Updated `VideoDto.kt` to remove unused import.
- Updated `TvShowDetailsRemoteMapper.kt` to use `fullVideoUrl` from `VideoDto`.
- Updated `AndroidManifest.xml` to simplify launcher activity declarations and remove the light/dark theme-specific aliases.
- Added logging in `VideoLauncher.kt` for debugging.

Noor-Serry committed:
Merge remote-tracking branch 'origin/feature/add-trailer-video-to-movie-and-tvshow' into feature/add-trailer-video-to-movie-and-tvshow

Noor-Serry committed:
Merge remote-tracking branch 'origin/feature/add-trailer-video-to-movie-and-tvshow' into feature/add-trailer-video-to-movie-and-tvshow

Noor-Serry committed:
Merge remote-tracking branch 'origin/feature/add-trailer-video-to-movie-and-tvshow' into feature/add-trailer-video-to-movie-and-tvshow

nabilBouzineDev commented on viewModel/src/main/java/com/amsterdam/viewmodel/seriesDetails/SeriesDetailsEffect.kt:
```diff
@@ -7,4 +7,5 @@ sealed interface SeriesDetailsEffect : BaseViewModel.BaseUiEffect {
     object NavigateToCastScreen : SeriesDetailsEffect
     object NavigateToLoginScreenEffect : SeriesDetailsEffect
     data class NavigateToSeriesDetails(val tvShowId: Long) : SeriesDetailsEffect
+    data class LaunchSeriesVideoEffect(val url: String) : SeriesDetailsEffect
```
This Name isn't consistent with the rest of names, why not change it to `NavigateToVideoTrailerSite`?

Noor-Serry replied:
I think navigate will indicate that we will navigatet to other screen in our app but its not our case here

saraashraf11 commented on repository/src/main/java/com/amsterdam/repository/dto/remote/VideoDto.kt:
```diff
@@ -0,0 +1,41 @@
+    val iso6391: String,
+
+    @SerialName("iso_3166_1")
+    val iso31661: String,
```
Can you rename this variable because it is not clear?

nabilBouzineDev commented on ui/src/main/java/com/amsterdam/ui/screens/VideoLauncher.kt:
```diff
@@ -0,0 +1,22 @@
+        videoUrl.toUri()
+    )
+
+    try {
```
Is there a scenario where user has no Youtube APP and No Browser? If yes, I think we should handle this case.

nabilBouzineDev commented on repository/src/main/java/com/amsterdam/repository/utils/VideoBaseUrl.kt:
```diff
@@ -0,0 +1,5 @@
+package com.amsterdam.repository.utils
+
+object VideoBaseUrl {
```
I think it'll be better if you move it to the VideoLauncher file if it's used only in one place?

Noor-Serry committed:
fix: handle video launch errors and refactor VideoDto

This commit addresses issues related to launching videos and refactors the `VideoDto`.

**Video Launch Error Handling:**
- Implemented error handling in `openYouTubeVideo` to catch general exceptions during video launch attempts.
- Added an `onLaunchFailed` callback to `openYouTubeVideo` to notify the caller of launch failures.
- In `MovieDetailsScreen` and `SeriesDetailsScreen`, if video launch fails, a SnackBar message "error to launch the video" (localized) is displayed.

**VideoDto Refactor:**
- Moved `YOUTUBE_BASE_URL` from the deleted `VideoBaseUrl.kt` object into a private companion object within `VideoDto.kt`.
- Renamed fields in `VideoDto` for clarity:
    - `iso6391` is now `languageCode`.
    - `iso31661` is now `countryCode`.

**Other Changes:**
- Added a new string resource `video_launch_error` and its Arabic translation.
- Added `isVideoLauncherFailed` to `MovieDetailsUiState` (though it appears unused in the provided diff).

eslam-99 commented on repository/src/main/java/com/amsterdam/repository/dto/local/LocalMovieDto.kt:
Don't you need to save video URL?

Noor-Serry replied:
no we do not need to cache it

Noor-Serry committed:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/add-trailer-video-to-movie-and-tvshow

# Conflicts:
#	repository/src/main/java/com/amsterdam/repository/mapper/local/MovieWithCategoriesLocalMapper.kt

saraashraf11 commented on entity/src/main/java/com/amsterdam/entity/Movie.kt:
```diff
@@ -14,6 +14,6 @@ data class Movie(
     val originCountry: String,
     val runTimeInMinutes: Int,
-    val hasVideo : Boolean,
+    val videoUrl: String = "",
```
It's recommended to use nullable types only when the data may genuinely be missing (e.g., videoUrl). Avoid using default empty values like "" or 0 to maintain data integrity

Noor-Serry replied:
I do not think that . actually I think the opposite

eslam-99 commented on ui/src/main/res/values-ar/string.xml:
```diff
@@ -221,4 +221,6 @@
     <!-- Details Screen -->
     <string name="there_is_no_gallery">لا توجد صور</string>
     <string name="there_is_no_company_production">لا توجد شركات إنتاج</string>
+    <string name="video_launch_error">error to launch the video</string>
```
Need translation here

eslam-99 commented on ui/src/main/res/values/strings.xml:
```diff
@@ -223,4 +223,6 @@
     <string name="there_is_no_company_production">There Is No Company Production</string>
     <string name="there_is_no_seasons">It Seem Like There Is No Seasons Yet</string>
 
+    <string name="video_launch_error">error to launch the video</string>
```
"Error launching" or "Failed to launch"

Noor-Serry committed:
Refactor: Update video launch error message

The error message displayed when a video fails to launch has been updated for clarity in both English and Arabic.

- English: "error to launch the video" -> "Failed to play the video"
- Arabic: "error to launch the video" -> "خطأ في تشغيل الفيديو"
